Toy model sensitivities

ben18785 opened the pull request:
Addresses #298 by making it much easier to add toy model sensitivities (via the forward sensitivities method).

This involves the following:

- Adding methods to ``ToyModel`` class that handle the formation of the RHS of the outputs + sensitivities combined ODE
- Added sensitivities for Goodwin Oscillator and restructured those for Fitzhugh-Nagumo to handle new structure of toy models
- Added short codes to example notebooks to highlight sensitivity functionality of these models

ben18785 commented:
@martinjrobins Before I add tests, what are your thoughts on this restructuring? It should make adding sensitivities for the remaining toy models trivial (which I increasingly believe will be important to have in our paper, where we benchmark NUTS vs other samplers).

martinjrobins commented on pints/toy/_toy_classes.py:
```diff
@@ -61,3 +79,71 @@ def suggested_times(self):
+        outputs and sensitivities. This should be a vector of length
+        ``n_outputs`` + ``n_parameters``.
+        """
+        y = y_and_dydp[0:self.n_outputs()]
```
maybe can make this a bit more readable by using temp variables for n_outputs and n_parameters?
```python
n = self.n_outputs()
m = self.n_parameters()
```

ben18785 replied:
Done

ben18785 commented:
@martinjrobins -- sorry -- another prod, if happy, can you merge please? I don't have the power to (which is a good thing!).

MichaelClerx commented on pints/toy/_toy_classes.py:
```diff
@@ -46,6 +48,22 @@ class ToyModel(object):
+        """
+        raise NotImplementedError
+
+    def jacobian(self, y, t, p):
```
What's the rationale for having `jacobian` public and the others private?

ben18785 replied:
Jacobians are useful for other things (like some gradient-based optimisers and samplers), so thought I'd give them a public method.

MichaelClerx commented on pints/toy/_toy_classes.py:
```diff
@@ -46,6 +48,22 @@ class ToyModel(object):
+        """
+        raise NotImplementedError
+
+    def jacobian(self, y, t, p):
```
Some description of `y, t, p` ?

ben18785 replied:
Done

MichaelClerx commented on pints/toy/_toy_classes.py:
```diff
@@ -61,3 +79,73 @@ def suggested_times(self):
+        """ See :meth:`pints.ForwardModel.simulate()`. """
+        return self._simulate(parameters, times, False)
+
+    def _simulate(self, parameters, times, sensitivities):
```
This is all very useful and a Good Idea, but specific to ODE models and could be useful to users implementing new toy ODE models.

Should we have a subclass called e.g. `ODEToyModel` that has all these methods, possibly public, and has more documentation on how to use this functionality?

ben18785 replied:
True, yes, you may be right that it's better to create a new class (I was mistakenly thinking that we just had ODE toy models, whereas we have others too).

MichaelClerx replied:
What do you think @martinjrobins ?

ben18785 replied:
I'm resolving this as think we decided to have a ToyODEModel class

MichaelClerx commented on pints/toy/_toy_classes.py:
```diff
@@ -61,3 +64,134 @@ def suggested_times(self):
+        raise NotImplementedError
+
+
+class ToyODEModel(ToyModel):
```
This implements some of the `ForwardModelS1` methods, and so should extend that class?

Unless we want to split into toy ode models that do and don't do derivatives, but I'm happy to only have one that does ODEs where derivatives are available, as long as we make that clear in the docstring!

ben18785 replied:
I've now updated things as we discussed in the meeting, with a separate ToyODEModel class.

MichaelClerx commented:
Made a few small tweaks to the docstrings @ben18785 

One last thing: We have the names "parameters" and "times" in some parts of the ToyODEModel class, but also generally throughout pints. So should we use them here too?

The order of the arguments (y, t, p) makes a lot of sense to me too, but isn't consistent with the other methods in the class (parameters, times), so wondering about that as well...

Thoughts @ben18785 @martinjrobins ?

ben18785 commented:
@MichaelClerx Thanks for having a look -- I am still working on this so probably should have had it as a draft PR; sorry. Good point re: parameters and times. Let me take a look..

ben18785 commented:
@MichaelClerx This is ready to review now, please. I'd suggest we make the parameters thing a separate issue, if you feel strongly about it. Having looked, I agree it is a bit ungainly, but not something I think we should address here.

MichaelClerx commented on pints/tests/test_toy_fitzhugh_nagumo_model.py:
```diff
@@ -59,6 +60,16 @@ def test_values(self):
+        parameters = [0.2, 0.7, 2.8]
+        times_finer = np.linspace(0, 20, 500)
+        sols, sens = model.simulateS1(parameters, times_finer)
+        f = interp1d(times_finer, sens[:, 0][:, 2])
```
What's going on here???

ben18785 replied:
I've added comments to these -- to obtain the sensitivity at a predefined time it's easiest to interpolate the sensitivity series.

MichaelClerx replied:
I don't understand. Why don't you request evaluation at the times you want?

ben18785 replied:
I could but find it easier to do it this way as it means I don't have to keep tabs on which observation maps to which time -- feel free to change if you feel strongly though.

MichaelClerx replied:
Made it easier, but found a bug!

MichaelClerx commented on pints/tests/test_toy_goodwin_oscillator_model.py:
```diff
@@ -40,6 +41,48 @@ def test_values(self):
         self.assertAlmostEqual(values[100, 1], 0.1779547, places=6)
         self.assertAlmostEqual(values[100, 2], 2.6074527, places=6)
 
+    def test_sensitivity(self):
```
Could you add a comment in each of these? I don't get what the difference between test_sensitivity and test_sensitivities is

ben18785 replied:
Good point -- I've combined these into a single test method as it didn't make sense to keep these separate.

MichaelClerx commented on pints/toy/_goodwin_oscillator_model.py:
```diff
@@ -9,27 +9,26 @@
+    Parameters are :math`[k_2, k_3, m_1, m_2, m_3]`. The initial conditions
+    are hard-coded at ``[0.0054, 0.053, 1.93]``.
 
     Extends :class:`pints.ForwardModel`, :class:`pints.toy.ToyModel`.
```
All modified models will need to update this line of the docstring

ben18785 replied:
Good spot -- done!

MichaelClerx commented on pints/toy/_lotka_volterra_model.py:
```diff
@@ -38,13 +41,41 @@ class LotkaVolterraModel(pints.ForwardModel, ToyModel):
     References
     ----------
     .. [1] https://en.wikipedia.org/wiki/Lotka%E2%80%93Volterra_equations
```
This could just say `https://en.wikipedia.org/wiki/Lotka-Volterra_equations`

MichaelClerx replied:
Without the %E2% etc

ben18785 replied:
changed

MichaelClerx commented on pints/toy/_toy_classes.py:
```diff
@@ -61,3 +64,155 @@ def suggested_times(self):
         """
         raise NotImplementedError
 
+    def simulate(self, parameters, times):
```
This shouldn't be here, as it's already defined in `ForwardModel`

MichaelClerx replied:
(which this class doesn't inherit from)

ben18785 replied:
removed this

MichaelClerx commented on pints/toy/_toy_classes.py:
```diff
@@ -61,3 +64,155 @@ def suggested_times(self):
+            np.transpose(self._dfdp(y, t, p)))
+        return np.concatenate((dydt, d_dydp_dt.reshape(-1)))
+
+    def simulate(self, parameters, times):
```
This one at least is functionally necessary...
@martinjrobins what do you think?

At the moment we've got the models themselves extending both ToyODEModel and either ForwardModel or ForwardModelS1 (they get to choose, which is important!)

With ToyModel we can avoid having a simulate(), but for ToyODEModel it's kind of necessary that we do. I think the cleanest solution would be to split ToyODEModel into a derivative-free and S1 class, and have those extend ForwardModel and ForwardModelS1. But what we have now works just as well and is probably easier...

ben18785 replied:
I think let's avoid splitting things up as it's already quite a complicated class ecosystem.

MichaelClerx commented:
But looks great @ben18785 ! The model classes look much tidier & I like the bonus improvements. Just need to fix a few small issues and work out how best to do the inheritance

ben18785 commented:
@MichaelClerx I've addressed your comments (hopefully). I may just have a quick blitz to try and get the last (possible) sensitivities in there...will buzz you when more reviewing needs to be done!

ben18785 commented:
@MichaelClerx Can you have a look now? I decided it's probably best to keep the remaining toy model sensitivities as a separate PR as this one is quite big now.

MichaelClerx commented:
Have simplified the tests @ben18785 
And found a small bug!
The first point returned is always 0, regardless of whether you ask for t=0 to be included in the returned times.
Probably an issue in ToyODEModel itself, as all three models that were using interp1d now show it. Have added a specific test for this issue to the fitzhugh model

ben18785 commented:
@MichaelClerx Thanks. It looks like the sensitivity tests are failing now though...Could you have a look at this as Martin and I need these for looking at NUTS?

MichaelClerx commented:
@ben18785 that's right! The code in this PR can't handle the case where the initial time point isn't in the requested time points so this needs fixing!

ben18785 commented:
Good spot. It looks like this is a requirement for `odeint` in [Scipy](https://docs.scipy.org/doc/scipy-0.17.0/reference/generated/scipy.integrate.odeint.html). I.e. that the first time point corresponds to the value of `y0`. As such, this will be an issue for ode integration in general - i.e. for the `simulate` as well as `simulateS1`, no?

In that case, I'd suggest we make a separate issue/PR to handle this corner case in general (and adapt the tests for this PR so that they pass) as it'll take some work to fix this. What do you think?

MichaelClerx commented:
I'd rather not merge in code that has known bugs :-/
I've got some time now though, so let me have a go at fixing it...
